### DISCUSSION TITLE
Need suggestion on optimization of search for >1m articles.

### DISCUSSION THREAD

**Discussion by coolbq** [FreshRSS/FreshRSS#3864] [Q&A] [answered]
Hi, I'm astonished that FreshRSS could easily support more than 1 MILLION  articles with a 1CPU/2G server. I made up some crawlers to automatically push crawled articles to the FreshRSS. This is really GREAT! FreshRSS could easily handle the browsing request.

But the only problem I met with is when I tried to search intitle or keyword alone. It's a little too slow for the first time search of a key word which usually takes a minute or so. I thought indexing could help. So I tried to make index and FULLTEXT index of the title, and narrow down the search using "intitle:keyword date:2020/". And I set 1G of memory to innodb_buffer_pool_size. But it seems FreshRSS does't hit the index at all. The same speed of search as not indexed. 

After research, I find out that FreshRSS uses the "title LIKE". So I'm wonderring the following:
1, is it possible to change the FreshRSS to use "MATCH AGAINST" to hit the fulltext index of title? 
2, if 1 is yes, how ?
3, if 2 is no, what's your suggestion to optimize the searching speed ? If content is not possible, at least there should be a way to boost the title field, right? 

Added info: I use Chinese in the title and content. 

Thanks in advance. and hoped for your detailed help. 
Sincerely

**Comment by coolbq** [comment]
And another thought: will partition the db in mysql help ?

**Comment by Alkarex** [comment, marked as answer]
Hello,
Maybe you could have a look at what PostgreSQL can do for that, which should work with classic `LIKE` SQL requests:
https://www.postgresql.org/docs/current/textsearch-indexes.html

See related issue https://github.com/FreshRSS/FreshRSS/issues/1331 , where we should in particular document how to do it more precisely.
Feedback welcome.

**Reply by Alkarex** [reply]
P.S. We offer command line export/import from any of our supported database formats to another, through an intermediate SQLite

**Reply by coolbq** [reply]
Thanks very much!
I'll look into it and see what will happen.

**Reply by Alkarex** [reply]
To migrate to PostgreSQL, see our command line https://github.com/FreshRSS/FreshRSS/tree/edge/cli

## 1. Export from MySQL to SQLite
```sh
cd ./FreshRSS/
./cli/export-sqlite-for-user.php --user alice --filename /tmp/alice.bak.sqlite
```

## 2. Reinstall FreshRSS using PostgreSQL
Either locally, or using Docker ( https://hub.docker.com/_/postgres )

## 3. Import from SQLite to PostgreSQL
```sh
./cli/import-sqlite-for-user.php --user alice --force-overwrite --filename /tmp/alice.bak.sqlite
```

## 4. Add native full-text indexes in PostgreSQL
https://www.postgresql.org/docs/current/textsearch-indexes.html

**Reply by coolbq** [reply]
I've done as guided but met with some problems with the indexing.
1, after creating the user in FreshRSS, I managed to delete the content and title columns and add another content and title columns with type as tsvector as guided in the official textsearch-indexes guided. And create an index on the content and title. By this way, the PostgreSQL doesn't allow to change the order of the columns. 
2, Then I execute the import sqlite php file to import the data. Then I recheck the entry table to find the column types of the content and title changed back to default and the indexing disappeared.

**Reply by coolbq** [reply]
I tried to not use --force-overwrite option in the command, but it failed with error message:
`Error during SQLite copy of feeds!
If you would like to clear the user database first, use the option --force-overwrite`

**Reply by coolbq** [reply]
I also tried to modify the type of title and content in Create _entry table section in the file "install.sql.pgsql.php" to tsvector, but this time the sqlite import command failed with the same error message as not use --force-overwrite but I actually used. After revert back the types to default, the sqlite import works again. 

Now I'm stucked here. Don't know what to do next.

**Reply by coolbq** [reply]
I give up this way. 
Finally, I setup FreshRSS locally and made a crawler to sync the online Fresh articles to the local instance through the fever API. 
And then I give MySQL 8G memory to cache the whole database in the memory. Now it's lightening fast to search in FreshRSS.

**Comment by coolbq** [comment]
Is it possible to let FreshRSS to support [sphinx search engine](http://sphinxsearch.com/) ? 
Currently the search can't hit index of titles in MySQL, maybe we can speed it up by introducint sphinx in to the search.

**Reply by coolbq** [reply]
It seems tiny-tiny-rss has a plugin to support sphinx for a longer time. https://git.tt-rss.org/fox/tt-rss/wiki/SphinxSearch

**Reply by Alkarex** [reply]
Did you give a try to native full text indexing in PostgreSQL?

**Reply by coolbq** [reply]
PostgreSQL is a new to me. Thogh I tried to install it locally, and installed FreshRSS to store data in PostgreSQL, but didn't find an easy way to import the 1M articles to the new instance. 
I also checked the CLI info you told me, but didn't figure out how. 
The entry table is about 4Gb.

**Reply by coolbq** [reply]
I came across sphinx and it seems that it could help GREATLY BOOST the speed of search. Since MySQL is the default supported DB, I think if you could make FreshRSS originally support sphinx or elasticsearch or something alike, it will help alot for us especially when it comes to be over 1M or more articles. Currently, you don't have a perfect solution for big data except PostgreSQL. 

I'm not only using FreshRSS as a RSS hub and reader, but also do marketing research on it. It really helps alot.

**Reply by coolbq** [reply]
hhm, I just found [ZomboDB](https://www.zombodb.com/documentation/) which seems to make Postgre support ElasticSearch, this seems very promising to my need. I'll dive into this term and see what will come up.

**Reply by Alkarex** [reply]
Our current MySQL configuration is using compressed text, which is unsuitable for that. See my answer regarding PostgreSQL instead https://github.com/FreshRSS/FreshRSS/discussions/3864#discussioncomment-1413416

**Reply by coolbq** [reply]
Thank you Alkarex very much! Your guide is great. it helps alot.. I'll try and see.
